### DISCUSSION TITLE
Design pattern of concurrent async tasks with BackgroundTasks

### DISCUSSION THREAD

**Discussion by LuchiLucs** [fastapi/fastapi#12190] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, BackgroundTasks
# from fastapi_utils.tasks import repeat_every
from src.utils.logger_utils import get_logger
import asyncio
from enum import Enum

app = FastAPI()
logger = get_logger(__name__)

class BaseTask:
    def __init__(self, name, wait_duration: int):
        self.name = name
        self.running = False
        self.wait_duration = wait_duration

    async def __call__(self):
        while self.running:
            await self.run_task()
            logger.debug(f"{self.name} completed, waiting for {self.wait_duration} seconds...")
            await asyncio.sleep(self.wait_duration)

    async def run_task(self):
        raise NotImplementedError

    def start(self):
        self.running = True

    def stop(self):
        self.running = False

class TaskB(BaseTask):
    async def run_task(self):
        logger.debug(f"{self.name} running...")
        await asyncio.sleep(2)

class TaskC(BaseTask):
    async def run_task(self):
        logger.debug(f"{self.name} running...")
        await asyncio.sleep(5)

period_in_seconds = 10

class TaskList(str, Enum):
    TASK_A = "TASK_A"
    TASK_B = "TASK_B"
    TASK_C = "TASK_C"
    ALL = "ALL"

tasks = [
    TaskB(TaskList.TASK_B.value, period_in_seconds),
    TaskC(TaskList.TASK_C.value, period_in_seconds),
]


# @repeat_every(seconds=period_in_seconds, wait_first=True)
# async def periodic_tasks():
#     for task in tasks:
#         if task.running:
#             asyncio.create_task(task())

@app.post("/start-tasks/{task_names}")
async def start_task(task_names: TaskList, background_tasks: BackgroundTasks):
    task_names = task_names.value
    if task_names == TaskList.ALL.value:
        to_be_started = tasks
    else:
        to_be_started = [t for t in tasks if t.name == task_names]

    for task in to_be_started:
        task.start()
        background_tasks.add_task(task)
    return {"message": f"{[t.name for t in to_be_started]} started"}


@app.post("/stop-tasks/{task_names}")
async def stop_task(task_names: TaskList, background_tasks: BackgroundTasks):
    task_names = task_names.value
    if task_names == TaskList.ALL.value:
        to_be_stopped = tasks
    else:
        to_be_stopped = [t for t in tasks if t.name == task_names]

    for task in to_be_stopped:
        task.stop()
    return {"message": f"{to_be_stopped} started"}


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8085)
```

### Description

If I have a list of Callable Async tasks and I wish to run them in a background and also restart them after a duration period, how can I design this?

Using the framework FastAPI I tried to code this toy example but as far as I understood, the BackgroundTasks list is run not concurrently but in a FIFO order. How can I fix this example or design this alternatively in a simple and elegant manner?

I wish to replace some legacy code base where there were a huge codebase dedicated to define custome Thread/Daemon objects with background/periodic properties but I feel like it was not consistent and it was overcomplex too.

### Operating System

Linux

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

3.11.0

**Comment by LuchiLucs** [comment]
I modified the start API with the following code and it seems to integrate without errors my list of coroutines which needs to be run concurrently with the expected callable function that background_tasks expect by means of wrapping:

```
@app.post("/start-tasks/{task_names}")
async def start_task(task_names: TaskList, background_tasks: BackgroundTasks):
    task_names = task_names.value
    if task_names == TaskList.ALL.value:
        to_be_started = tasks
    else:
        to_be_started = [t for t in tasks if t.name == task_names]

    task_coroutines = []
    for task in to_be_started:
        task.start()
        task_coroutines.append(task())

    async def callable_run_tasks():
        await asyncio.gather(*task_coroutines, return_exceptions=True)

    background_tasks.add_task(callable_run_tasks)
    return {"message": f"{[t.name for t in to_be_started]} started"}
```

What do you think? I wish to discuss this design

**Reply by LuchiLucs** [reply]
Update: the provided toy solution seems not to working as expected as the server seems not able to serve other requests while the tasks are running: i.e. they are not so "background" tasks. Still have to investigate why, in the meanwhile I would appreciate inputs. Thanks!

**Comment by YuriiMotov** [comment]
`BackgroundTasks` is for simple use cases. If you need more, use distributed task queues like Celery, ARQ, etc..

Let's close this for now to keep the discussions clean.
If you still need help, feel free to let us know.
